lib/scanner/walk.go

```diff
@@ -248,43 +293,108 @@ func (w *walker) walkAndHashFiles(ctx context.Context, fchan, dchan chan protoco
 		}
 
 		if w.Matcher.Match(path).IsIgnored() {
+			if oldFile == nil {
+				return skip
+			}
+			finishedChan <- ScanResult{
+				FileInfo: oldFile.InvalidatedCopy(w.ShortID),
+				Old:      oldFile,
+			}
+			for haveChanOpen && strings.HasPrefix(currDBFile.Name, path+string(fs.PathSeparator)) {
+				if !currDBFile.IsInvalid() {
+					finishedChan <- ScanResult{
+						FileInfo: currDBFile.InvalidatedCopy(w.ShortID),
+						Old:      currDBFile,
+					}
+				}
+				currDBFile, haveChanOpen = <-w.HaveChan
+			}
```

Comment:
To be honest, perhaps it's usefull to change to to two channels, each walking it's own thing, which then pulls a value from each, if matches, does stuff, if not, handles addition/removal and  tugs off the other channel. atleast it would be less confusing.

Reply:
I don't immediately see how this would be any different from what is done right now: New path is coming from fs walk, then db file infos are received and handled until they catch up with the fs path, then magic happens including handling child db items if necessary. How would that be different with two channels? (I am not saying the current state isn't confusing)

Reply:
I would have more confidence in this working, debugging and extending this, as the lockstepping is now peppered throughout the file in multiple places.